src/builtins.ts

```diff
@@ -2056,6 +2072,267 @@ function builtin_store(ctx: BuiltinContext): ExpressionRef {
 }
 builtins.set(BuiltinNames.store, builtin_store);
 
+// add<T?>(left: T, right: T) -> T
+function builtin_add(ctx: BuiltinContext): ExpressionRef {
+  var compiler = ctx.compiler;
+  var module = compiler.module;
+  if (checkTypeOptional(ctx, true) | checkArgsRequired(ctx, 2))
+    return module.unreachable();
+  var operands = ctx.operands;
+  var typeArguments = ctx.typeArguments;
+  var left = operands[0];
+  var arg0 = typeArguments
+    ? compiler.compileExpression(
+        left,
+        typeArguments[0],
+        Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+      )
+    : compiler.compileExpression(operands[0], Type.auto, Constraints.MUST_WRAP);
+  var type = compiler.currentType;
+  if (type.isValue) {
+    let arg1: ExpressionRef;
+    if (!typeArguments && left.isNumericLiteral) {
+      // prefer right type
+      arg1 = compiler.compileExpression(
+        operands[1],
+        type,
+        Constraints.MUST_WRAP
+      );
+      if (compiler.currentType != type) {
+        arg0 = compiler.compileExpression(
+          left,
+          (type = compiler.currentType),
+          Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+        );
+      }
+    } else {
+      arg1 = compiler.compileExpression(
+        operands[1],
+        type,
+        Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+      );
+    }
+    let op: BinaryOp = -1;
+    switch (type.kind) {
+      case TypeKind.I8:
+      case TypeKind.I16:
+      case TypeKind.I32:
+      case TypeKind.U8:
+      case TypeKind.U16:
+      case TypeKind.U32:
+      case TypeKind.BOOL: {
+        return compiler.ensureSmallIntegerWrap(
+          module.binary(BinaryOp.AddI32, arg0, arg1),
+          type
+        );
+      }
```

Comment:
```suggestion
      case TypeKind.I8:
      case TypeKind.U8:
      case TypeKind.I16:
      case TypeKind.U16:
      case TypeKind.BOOL: {
        return compiler.ensureSmallIntegerWrap(
          module.binary(BinaryOp.AddI32, arg0, arg1),
          type
        );
      }
      case TypeKind.I32:
      case TypeKind.U32: {
        op = BinaryOp.AddI32;
        break;
      }
```
Use `ensureSmallIntegerWrap` for `TypeKind.I32` and `TypeKind.U32` is unnecessary

Reply:
Changed for add, mul & sub